### DISCUSSION TITLE
Render speed question in "HttpKernelRunner" and "fastcgi_finish_request" for Symfony 6.4+ on dev machine

### DISCUSSION THREAD

**Discussion by kdobryansky** [symfony/symfony#57303] [Q&A]
Intro. I'm using win10 with VMWare player, Ubuntu20 server inside with nginx and php8.2-fpm

After upgrading Symfony from 6.3 to 6.4 (or 7+) I have found significant speed degradation. Every page renders about 8 seconds (even blank controller page with empty `Response` class and no business logic), but symfony profiler shows only 80-100ms.

After some investigation, i found that problem is in "`run`" method of `HttpKernelRunner` class (Here is pull request for this change https://github.com/symfony/symfony/pull/52079). In version below 6.4 it has "`$response->send()`" call with default true argument which inside caused calling "`fastcgi_finish_request`" and everything worked fast.
But in 6.4+ version, "`$response->send(false)`" method now getting false value and it causes return to the parent method before "`fastcgi_finish_request`" is called. And in "`run`" it can call "`fastcgi_finish_request`" only with `APP_DEBUG=0` value (debug is disabled) in env file which is always enabled on local dev machine. So it skips this call and go forward to "`Response::closeOutputBuffers`" and "`flush`" call.

So, the question is, what is the purpose for "`&& !$this->debug`" which is causing not calling "`fastcgi_finish_request`"? Before 6.4 debug state didn't matter and symfony always called "`fastcgi_finish_request`" and everything went fine.
And what benefits do we have now in 6.4+ by calling following code with debug enabled instead of "`fastcgi_finish_request`": 
```php
Response::closeOutputBuffers(0, true);
flush();
``` 

Thank you in advance. I'm really get stuck here.

Here is the code of run method to have a picture of it:

```php
public function run(): int
{
	$response = $this->kernel->handle($this->request);

	if (Kernel::VERSION_ID >= 60400) {
		$response->send(false);

		if (\function_exists('fastcgi_finish_request') && !$this->debug) {
			fastcgi_finish_request();
		} elseif (\function_exists('litespeed_finish_request') && !$this->debug) {
			litespeed_finish_request();
		} else {
			Response::closeOutputBuffers(0, true);
			flush();
		}
	} else {
		$response->send();
	}

	if ($this->kernel instanceof TerminableInterface) {
		$this->kernel->terminate($this->request, $response);
	}

	return 0;
}
```

**Comment by KostyaKulakov** [comment]
The purpose of `&& !$this->debug` condition in the run method of the HttpKernelRunner class is to prevent calling fastcgi_finish_request when the debug mode is enabled. This is because `fastcgi_finish_request` is a function that is used to terminate the request processing and send the response to the client. When the debug mode is enabled, Symfony needs to perform additional tasks, such as collecting debug information and displaying the profiler, before sending the response to the client. Calling fastcgi_finish_request too early would prevent Symfony from performing these tasks.

The benefits of calling `Response::closeOutputBuffers(0, true);` and `flush();` instead of `fastcgi_finish_request` when the debug mode is enabled are:

**It allows Symfony to perform additional tasks, such as collecting debug information and displaying the profiler, before sending the response to the client.**
 It ensures that all output buffers are flushed before the response is sent to the client. This can help to prevent problems such as headers being sent after the response body has already been sent.

In general, it is recommended to leave the `&& !$this->debug` condition in the run method of HttpKernelRunner class as it is. This will ensure that Symfony can perform all of the necessary tasks before sending the response to the client.

**Comment by phpinfo** [comment]
I’m going through the exact same thing!
